[State Mgmt] Docset refresh (#2541)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-stateful-service.md

description: "Use state management with a scaled, replicated service"
---

In this article, you'll learn how to create a stateful service which can be horizontally scaled, using opt-in concurrency and consistency models. Consuming the state management API frees developers from difficult state coordination, conflict resolution, and failure handling.

## Set up a state store

A state store component represents a resource that Dapr uses to communicate with a database.
For the purpose of this guide, we'll use the default Redis state store.

### Using the Dapr CLI

> When you run `dapr init` in standalone mode, Dapr creates a default Redis `statestore.yaml` and runs a Redis state store on your local machine, located:

Reviewer: ```suggestion
When you run `dapr init` in self-hosted mode, Dapr creates a default Redis `statestore.yaml` and runs a Redis state store on your local machine, located:
```

- On Windows, under `%UserProfile%\.dapr\components\statestore.yaml`
- On Linux/MacOS, under `~/.dapr/components/statestore.yaml`

With the `statestore.yaml` component, you can easily swap out underlying components without application code changes.

See a [list of supported state stores]({{< ref supported-state-stores >}}).

### Kubernetes

See [how to setup different state stores on Kubernetes]({{<ref setup-state-store>}}).

## Strong and eventual consistency

> Using strong consistency, Dapr will make sure the underlying state store:

Reviewer: ```suggestion
Using strong consistency, Dapr makes sure that the underlying state store:
```

> - Returns the response once the data has been written to all replicas

Reviewer: ```suggestion
- Returns the response once the data has been written to all replicas.
```

> - Receives an ack from a quorum before writing or deleting state

Reviewer: ```suggestion
- Receives an ACK from a quorum before writing or deleting state.
```

> For get requests, Dapr will make sure the store returns the most up-to-date data consistently among replicas. The default is eventual consistency, unless specified otherwise in the request to the state API.

Reviewer: ```suggestion
For get requests, Dapr makes sure the store returns the most up-to-date data consistently among replicas. The default is eventual consistency, unless specified otherwise in the request to the state API.
```

response = requests.delete(dapr_state_url + "/key1", headers={"consistency":"strong"})
```

If the `concurrency` option hasn't been specified, the default is last-write concurrency mode.

## First-write-wins and last-write-wins

Dapr allows developers to opt-in for two common concurrency patterns when working with data stores: 

- **First-write-wins**: useful in situations where you have multiple instances of an application, all writing to the same key concurrently.
- **Last-write-wins**: Default mode for Dapr.

Dapr uses version numbers to determine whether a specific key has been updated. You can:

1. Retain the version number when reading the data for a key.
1. Use the version number during updates such as writes and deletes. 

> If the version information has changed since the version number was retrieved, an error is thrown, requiring the you to perform another read to get the latest version information and state.

Reviewer: ```suggestion
If the version information has changed since the version number was retrieved, an error is thrown, requiring you to perform another read to get the latest version information and state.
```

> Dapr utilizes ETags to determine the state's version number. ETags are returned from state requests in an `ETag` header. Using ETags, you know that a resource has been updated since the last time they checked by erroring during an ETag mismatch.

Reviewer: ```suggestion
Dapr utilizes ETags to determine the state's version number. ETags are returned from state requests in an `ETag` header. Using ETags, your application knows that a resource has been updated since the last time they checked by erroring during an ETag mismatch.
```

The following example shows how to:

- Get an ETag

> - Use the ETag to save state

Reviewer: ```suggestion
- Use the ETag to save state.
```

> - Delete the state

Reviewer: ```suggestion
- Delete the state.
```
